data_model/src/isi.rs

```diff
@@ -4,865 +4,953 @@
-        Self {
-            object_id: object_id.into(),
-            key: key.into(),
-        }
+        Self::TriggerRepetitions(Mint::new(repetitions, trigger_id))
     }
 }
 
-impl RegisterExpr {
-    /// Construct [`Register`].
-    pub fn new<O: Into<EvaluatesTo<RegistrableBox>>>(object: O) -> Self {
-        Self {
-            object: object.into(),
-        }
+isi_box! {
+    /// Enum with all supported [`Burn`] instructions.
+    pub enum BurnBox {
+        /// Burn [`PublicKey`] for [`Account`].
+        AccountPublicKey(Burn<PublicKey, Account>),
+        /// Burn [`Asset`].
+        Asset(AssetBurnBox),
+        /// Burn [`Trigger`] repetitions.
+        TriggerRepetitions(Burn<u32, Trigger<TriggeringFilterBox>>),
     }
 }
 
-impl UnregisterExpr {
-    /// Construct [`Unregister`].
-    pub fn new<O: Into<EvaluatesTo<IdBox>>>(object_id: O) -> Self {
-        Self {
-            object_id: object_id.into(),
-        }
+isi_box! {
+    /// Enum with all supported [`Burn`] instructions related to [`Asset`].
+    pub enum AssetBurnBox {
+        /// Burn [`Asset`] of [`Quantity`] type.
+        Quantity(Burn<u32, Asset>),
+        /// Burn [`Asset`] of [`BigQuantity`] type.
+        BigQuantity(Burn<u128, Asset>),
+        /// Burn [`Asset`] of [`Fixed`] type.
+        Fixed(Burn<Fixed, Asset>),
     }
 }
 
-impl MintExpr {
-    /// Construct [`Mint`].
-    pub fn new<O: Into<EvaluatesTo<Value>>, D: Into<EvaluatesTo<IdBox>>>(
-        object: O,
-        destination_id: D,
+impl BurnBox {
+    /// Constructs a new [`BurnBox`] to burn [`PublicKey`] for [`Account`].
+    pub fn account_public_key(
+        public_key: PublicKey,
+        account_id: <Account as Identifiable>::Id,
     ) -> Self {
-        Self {
-            object: object.into(),
-            destination_id: destination_id.into(),
-        }
+        Self::AccountPublicKey(Burn::new(public_key, account_id))
     }
-}
 
-impl BurnExpr {
-    /// Construct [`Burn`].
-    pub fn new<O: Into<EvaluatesTo<Value>>, D: Into<EvaluatesTo<IdBox>>>(
-        object: O,
-        destination_id: D,
-    ) -> Self {
-        Self {
-            object: object.into(),
-            destination_id: destination_id.into(),
-        }
+    /// Constructs a new [`BurnBox`] to burn [`Asset`] of [`Quantity`] type.
+    pub fn asset_quantity(quantity: u32, asset_id: <Asset as Identifiable>::Id) -> Self {
+        Self::Asset(AssetBurnBox::Quantity(Burn::new(quantity, asset_id)))
+    }
+
+    /// Constructs a new [`BurnBox`] to burn [`Asset`] of [`BigQuantity`] type.
+    pub fn asset_big_quantity(quantity: u128, asset_id: <Asset as Identifiable>::Id) -> Self {
+        Self::Asset(AssetBurnBox::BigQuantity(Burn::new(quantity, asset_id)))
+    }
+
+    /// Constructs a new [`BurnBox`] to burn [`Asset`] of [`Fixed`] type.
+    pub fn asset_fixed(quantity: Fixed, asset_id: <Asset as Identifiable>::Id) -> Self {
+        Self::Asset(AssetBurnBox::Fixed(Burn::new(quantity, asset_id)))
     }
```

Comment:
```suggestion
    /// Constructs a new [`BurnBox`] to burn [`Asset`]
    pub fn asset(quantity: impl Into<AssetType>, asset_id: <Asset as Identifiable>::Id) -> Self {
        Self::Asset(AssetBurnBox::Quantity(Burn::new(quantity, asset_id)))
    }
```

is this not possible?

Reply:
ok, I see now. Can we do it like this:

```suggestion
    /// Constructs a new [`BurnBox`] to burn [`Asset`]
    pub fn asset_quantity(quantity: impl Into<AssetValue>, asset_id: <Asset as Identifiable>::Id) -> Self {
        match quantity.into() {
                AssetType::Quantity(quantity) => Self::Asset(AssetBurnBox::Quantity(Burn::new(quantity, asset_id))),
                AssetType::BigQuantity(quantity) => Self::Asset(AssetBurnBox::Quantity(Burn::new(quantity, asset_id))),
                AssetType::Fixed(quantity) => Self::Asset(AssetBurnBox::Quantity(Burn::new(quantity, asset_id))),
        }
    }
```

this is how I would prefer doing it